File: gui_view/src/main/java/com/goexp/galgame/gui/view/game/listview/imglist/CellController.java
package com.goexp.galgame.gui.view.game.listview.imglist;

import com.goexp.common.util.DateUtil;
import com.goexp.galgame.common.model.GameState;
import com.goexp.galgame.gui.model.Game;
import com.goexp.galgame.gui.util.Tags;
import com.goexp.galgame.gui.util.res.Images;
import com.goexp.galgame.gui.util.res.LocalRes;
import com.goexp.galgame.gui.view.DefaultController;
import com.goexp.galgame.gui.view.game.HomeController;
import com.goexp.galgame.gui.view.game.part.StateChangeController;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.effect.ColorAdjust;
import javafx.scene.image.ImageView;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.FlowPane;
import javafx.scene.layout.HBox;
import javafx.scene.text.Text;

public class CellController extends DefaultController {

    private Game game;

    @FXML
    private StateChangeController changeStateController;

    @FXML
    private ImageView imageImg;

    @FXML
    private HBox boxStar;

    @FXML
    private Text txtName;

    @FXML
    private Text txtSubName;

    @FXML
    private Label lbDate;

    @FXML
    private Label lbBrand;

    @FXML
    private FlowPane flowTag;


    protected void initialize() {


    }


    public void load(Game game) {
        this.game = game;

        txtName.setText(game.getMainName());
        txtSubName.setText(game.getSubName());

        lbBrand.setText(game.brand.name());
        lbDate.setText(DateUtil.formatDate(game.publishDate));
        changeStateController.load(game);

        if (game.tag.size() > 0)
            flowTag.getChildren().setAll(Tags.toNodes(game.tag));


        if (game.isOkImg()) {

            imageImg.setImage(Images.GameImage.small(game));

            if (game.state.get() == GameState.BLOCK)
                imageImg.setEffect(new ColorAdjust(0, -1, 0, 0));
            else
                imageImg.setEffect(null);
        } else {
            imageImg.setImage(null);
        }

        var image = LocalRes.HEART_32_PNG.get();
        boxStar.getChildren().clear();
        for (var i = 0; i < game.star; i++) {
            boxStar.getChildren().add(new ImageView(image));
        }

    }

//    public void link_OnAction(ActionEvent actionEvent) {
//        System.out.println("go");
//
//    }

    public void onClick(MouseEvent mouseEvent) {
        HomeController.$this.loadDetail(game);
    }

}

File: gui_view/src/main/scala/com/goexp/galgame/gui/view/game/listview/imglist/CellController.scala
package com.goexp.galgame.gui.view.game.listview.imglist

import com.goexp.common.util.date.DateUtil
import com.goexp.galgame.common.model.game.CommonGame.Titles
import com.goexp.galgame.common.model.game.GameState
import com.goexp.galgame.gui.model.Game
import com.goexp.galgame.gui.util.Tags
import com.goexp.galgame.gui.util.res.LocalRes
import com.goexp.galgame.gui.util.res.gameimg.GameImage
import com.goexp.galgame.gui.view.game.part.StateChangeController
import com.goexp.galgame.gui.view.{DefaultController, MainController}
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.effect.ColorAdjust
import javafx.scene.image.ImageView
import javafx.scene.input.MouseEvent
import javafx.scene.layout.{FlowPane, HBox}
import javafx.scene.text.Text

class CellController extends DefaultController {
  private var game: Game = _
  @FXML private var changeStateController: StateChangeController = _
  @FXML private var imageImg: ImageView = _
  @FXML private var boxStar: HBox = _
  @FXML private var txtName: Text = _
  @FXML private var txtSubName: Text = _
  @FXML private var lbDate: Label = _
  @FXML private var lbBrand: Label = _
  @FXML private var flowTag: FlowPane = _

  override protected def initialize() = {
  }

  def load(game: Game) = {
    this.game = game

    val Titles(mainTitle, subTitle) = game.getTitles
    txtName.setText(mainTitle)
    txtSubName.setText(subTitle)

    lbBrand.setText(game.brand.name)
    lbDate.setText(DateUtil.formatDate(game.publishDate))
    changeStateController.load(game)
    if (game.tag.size > 0) {
      val nodes = Tags.toNodes(game.tag) { tag =>
        val tagLabel = new Label(tag)
        tagLabel.getStyleClass.add("tag")
        tagLabel.getStyleClass.add("tagsmall")
        tagLabel
      }

      flowTag.getChildren.setAll(nodes)
    }

    imageImg.setImage {
      if (game.isOkImg) new GameImage(game).normal() else null
    }

    if (game.isOkImg) {
      imageImg.setEffect {
        if (game.state.get eq GameState.BLOCK) new ColorAdjust(0, -1, 0, 0) else null
      }
    }

    val image = LocalRes.HEART_16_PNG

    boxStar.getChildren.clear()
    val stars = Range(0, game.star).to(LazyList).map { _ => new ImageView(image) }.toArray
    boxStar.getChildren.addAll(stars: _*)

  }

  def onClick(mouseEvent: MouseEvent) =
    MainController().loadDetail(game)
}